[AWSIC] Update SCIM token with `tctl`

tcsc opened the pull request:
Implements SCIM token rotation for AWS IC, both to be generally useful and as an example of how to use the Update Creds API for other plugins. 

Usage:
```
$ tctl plugins update-creds awsic ${TOKEN}
```

Requires: #57690
Requires: gravitational/teleport.e#7016
Changelog: Added the ability to update the AWS Identity Center SCIM token in `tctl`

flyinghermit commented on tool/tctl/common/plugin/awsic.go:
```diff
@@ -316,3 +317,58 @@ func (p *PluginsCommand) InstallAWSIC(ctx context.Context, args installPluginArg
+		},
+	}
+
+	_, err = args.plugins.UpdatePluginStaticCredentials(ctx, &req)
```
Do we validate credential before it gets to the backend?

tcsc replied:
No, but I'm sure we _could_.

tcsc replied:
...and I had to revert it again because it requires packages in `teleport.e`

flyinghermit replied:
Hmm, any reason we can't move that to OSS? Validating that anywhere before it gets to the storage backend would do. We have had a couple of support ticket related to failed ic sync due to invalid scim credential. It would be helpful if we can catch such token before its saved in Teleport.

tcsc replied:
Moved `lib/scim/sdk` to OSS.

zmb3 commented on tool/tctl/common/plugin/plugins_command.go:
```diff
@@ -169,6 +176,12 @@ func (p *PluginsCommand) Cleanup(ctx context.Context, args installPluginArgs) er
 }
 
+func (p *PluginsCommand) initUpdateCreds(parent *kingpin.CmdClause) {
+	p.updateCreds.cmd = parent.Command("update-creds", "Updates a plugin's credentials.")
```
Sorry, now that I think about it we have `tctl auth rotate` for rotating CAs, would `tctl plugins rotate` make more sense for plugin creds?

tcsc replied:
No worries - either name is fine as long as its consistent.

zmb3 commented on tool/tctl/common/plugin/awsic.go:
```diff
@@ -316,3 +317,58 @@ func (p *PluginsCommand) InstallAWSIC(ctx context.Context, args installPluginArg
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
```
Do we need to make sure that the `plugin` that was returned is an AWSIC plugin? What happens if you run this command and pass it the name of some other plugin?

tcsc committed:
[AWSIC] Rotate IC ntegration SCIM token with tctl

Implements SCIM token rotation for AWS IC, both to be generally useful and as
an example of how to use the Update Creds API for other plugins.

Usage: `$ tctl plugins update-creds awsic ${TOKEN} `

Requires: #57690
Requires: gravitational/teleport.e#7016
Changelog: Added the ability to update the AWS Identity Center SCIM token in tctl

tcsc committed:
Reverted SCIM validation as it relies on `teleport.e`

flyinghermit commented on tool/tctl/common/plugin/awsic_test.go:
```diff
@@ -315,3 +320,202 @@ func TestUseSystemCredentialsInput(t *testing.T) {
+			test.assertError(t, err)
+
+			pluginsClient.AssertExpectations(t)
+			require.Equal(t, test.expectValidation, scimClient.pingCalled,
+				"SCIM validation Ping expected: %t, Ping called %t", test.expectValidation, scimClient.pingCalled)
```
I am confused, you mentioned that validation is now reverted but we have test here ? 
https://github.com/gravitational/teleport/pull/57691#discussion_r2321133429

tcsc committed:
Merge remote-tracking branch 'origin/master' into tcsc/tctl-rotate-creds

tcsc committed:
Fix tests

tcsc commented:
> Validating that anywhere before it gets to the storage backend would do.

We could probably do it in the gRPC service.

tcsc committed:
Moves SCIM SDK to OSS, restores token validation

tcsc committed:
fix license

flyinghermit commented on lib/scim/sdk/client.go:
```diff
@@ -0,0 +1,643 @@
+	return &listResp, nil
+}
+
+// Ping checks the connection to the SCIM server by sending a request to the ServiceProviderConfig endpoint.
+func (c *client) Ping(ctx context.Context) error {
+	u, err := c.endpointURL("ServiceProviderConfig")
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	resp, err := c.do(ctx, u, http.MethodGet, nil)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer resp.Body.Close()
+	switch resp.StatusCode {
+	case http.StatusOK:
+	case http.StatusUnauthorized, http.StatusForbidden:
+		return trace.AccessDenied("unauthorized")
+	case http.StatusInternalServerError:
+		return trace.BadParameter("internal server error")
+	default:
+		return trace.BadParameter("unexpected status code %v", resp.StatusCode)
+	}
+	return nil
+}
```
Instead of moving the whole scim sdk to oss, can't we just moved this logic to oss? Convert this to a function that accepts a ready made http client param and we have a common validator that can be used in both tctl and web ui?

tcsc replied:
Seems reasonable, done.

tcsc committed:
Replace SCIM SDK with simple HTTP ping

tcsc committed:
remove duplicate import

tcsc committed:
lint fixup

tcsc committed:
Merge branch 'master' into tcsc/tctl-rotate-creds

tcsc committed:
Merge branch 'master' into tcsc/tctl-rotate-creds

tcsc committed:
Merge branch 'master' into tcsc/tctl-rotate-creds

tcsc committed:
Merge branch 'master' into tcsc/tctl-rotate-creds
